src/rust/engine/options/src/lib.rs

```diff
@@ -297,86 +304,129 @@ impl OptionParser {
         }
         Ok(OptionValue {
             source: Source::Default,
-            value: default,
+            value: default.to_owned(),
         })
     }
 
+    pub fn parse_bool(&self, id: &OptionId, default: bool) -> Result<OptionValue<bool>, String> {
+        self.parse_scalar(id, &default, |source, id| source.get_bool(id))
+    }
+
     pub fn parse_int(&self, id: &OptionId, default: i64) -> Result<OptionValue<i64>, String> {
-        for (source_type, source) in self.sources.iter() {
-            if let Some(value) = source.get_int(id)? {
-                return Ok(OptionValue {
-                    source: *source_type,
-                    value,
-                });
-            }
-        }
-        Ok(OptionValue {
-            source: Source::Default,
-            value: default,
-        })
+        self.parse_scalar(id, &default, |source, id| source.get_int(id))
     }
 
     pub fn parse_float(&self, id: &OptionId, default: f64) -> Result<OptionValue<f64>, String> {
-        for (source_type, source) in self.sources.iter() {
-            if let Some(value) = source.get_float(id)? {
-                return Ok(OptionValue {
-                    source: *source_type,
-                    value,
-                });
-            }
-        }
-        Ok(OptionValue {
-            source: Source::Default,
-            value: default,
-        })
+        self.parse_scalar(id, &default, |source, id| source.get_float(id))
     }
 
     pub fn parse_string(
         &self,
         id: &OptionId,
         default: &str,
     ) -> Result<OptionValue<String>, String> {
-        for (source_type, source) in self.sources.iter() {
-            if let Some(value) = source.get_string(id)? {
-                return Ok(OptionValue {
-                    source: *source_type,
-                    value,
-                });
+        self.parse_scalar(id, default, |source, id| source.get_string(id))
+    }
+
+    #[allow(clippy::type_complexity)]
+    fn parse_list<T>(
+        &self,
+        id: &OptionId,
+        default: Vec<T>,
+        getter: fn(&Rc<dyn OptionsSource>, &OptionId) -> Result<Option<Vec<ListEdit<T>>>, String>,
+        remover: fn(Vec<T>, &Vec<T>) -> Vec<T>,
+    ) -> Result<Vec<T>, String> {
+        let mut list_edits: Vec<ListEdit<T>> = vec![];
+        for (_, source) in self.sources.iter() {
+            if let Some(edits) = getter(source, id)? {
+                list_edits.extend(edits);
```

Comment:
However, when reading the code... Shouldn't this look at settings the reverse order? Won't this apply the highest-priority edits first, since they'll occur earlier in the sources map? And defaults come last.

Reply:
Yeeesh, I believe you're right! This appears to be a longstanding bug that I guess was never noticed because we only used parse_string_list for fingerprinting bootstrap options... Great catch!

Reply:
OK, I wrote some tests to verify that the old implementation was wrong, as was my change, and I've fixed it. But while doing so I've discovered another issue with how the old implementation handles multiple config files, so this is going to take a little more work. I'll ping here when it's ready for another look.